### PR TITLE
BF16 end to end test using tpp-run

### PR DESCRIPTION
**Pull request by KavithaTipturMadhu**

### CODE REVIEW

File: test/BF16/mlp-all-bf16-tpprun.mlir [hunk 1]
```diff
@@ -0,0 +1,42 @@
+// RUN: tpp-opt %s linalg-ext-to-loops -convert-linalg-to-loops -convert-tpp-to-xsmm -convert-xsmm-to-func -convert-vector-to-scf -convert-scf-to-cf -lower-affine  -reconcile-unrealized-casts |\
```
Comment (chelini) [inline test/BF16/mlp-all-bf16-tpprun.mlir:1]: something is off here: need the `-` (linalg-ext-to-loops)

File: test/BF16/mlp-all-bf16-tpprun.mlir [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+    tpp.vnni_matmul ins(%arg11 : memref<128x1024xbf16>, %relayout_arg11 : memref<512x2048x2xbf16>) out(%arg10 : memref<128x2048xbf16>)
+    tpp.relu out(%arg10 : memref<128x2048xbf16>)
+
+    tpp.identity ins(%arg8 : memref<1000xbf16>) out(%arg9 : memref<128x1000xbf16>)
+    %relayout_arg10 = memref.alloc():memref<1024x1000x2xbf16>
+    %3 = memref.get_global @arg7:memref<2048x1000xbf16>
+    linalgx.pack %3 inner_dims_pos = [0] inner_tiles = [2] into %relayout_arg10:(memref<2048x1000xbf16> memref<1024x1000x2xbf16>)
+    tpp.vnni_matmul ins(%arg10 : memref<128x2048xbf16>, %relayout_arg10 : memref<1024x1000x2xbf16>) out(%arg9 : memref<128x1000xbf16>)
+    tpp.relu out(%arg9 : memref<128x1000xbf16>)
+
```
Comment (chelini) [inline test/BF16/mlp-all-bf16-tpprun.mlir:40]: should we add a check here?

File: tpp-run/MLIRBench.cpp [hunk 3]
```diff
@@ -213,7 +216,18 @@ Value MLIRBench::getTimerStats(Value acc) {
...
+  auto op = vector;
+  if (vector.getType().dyn_cast<VectorType>().getElementType().isBF16()) {
+    SmallVector<int64_t> dims;
+    for (int i = 0; i < vector.getType().dyn_cast<VectorType>().getRank();
+         i++) {
+      dims.push_back(vector.getType().dyn_cast<VectorType>().getShape()[i]);
+    }
+    VectorType vecType = VectorType::get(dims, builder.getF32Type());
+    op = builder.create<arith::ExtFOp>(unkLoc, vecType, vector, std::nullopt);
+    op.dump();
```
Comment (chelini) [inline tpp-run/MLIRBench.cpp:228]: debug

File: tpp-run/MLIRBench.cpp [hunk 4]
```diff
@@ -213,7 +216,18 @@ Value MLIRBench::getTimerStats(Value acc) {
 }
 
 void MLIRBench::printVector(Value vector) {
-  builder.create<vector::PrintOp>(unkLoc, vector);
+  auto op = vector;
+  if (vector.getType().dyn_cast<VectorType>().getElementType().isBF16()) {
```
Comment (rengolin) [inline tpp-run/MLIRBench.cpp:220]: repetition of `getType().dyn_cast<VectorType>()`, cache in a local variable.

File: tpp-run/MLIRBench.cpp [hunk 5]
```diff
@@ -213,7 +216,18 @@ Value MLIRBench::getTimerStats(Value acc) {
...
+    SmallVector<int64_t> dims;
+    for (int i = 0; i < vector.getType().dyn_cast<VectorType>().getRank();
+         i++) {
+      dims.push_back(vector.getType().dyn_cast<VectorType>().getShape()[i]);
+    }
+    VectorType vecType = VectorType::get(dims, builder.getF32Type());
```
Comment (rengolin) [inline tpp-run/MLIRBench.cpp:226]: can you not just `get( getShape(), F32 )`?

File: tpp-run/MLIRBench.cpp [hunk 6]
```diff
@@ -331,7 +359,8 @@ llvm::StringRef MLIRBench::createGlobal(MemRefType type) {
   auto tensorType =
       RankedTensorType::get(memrefTy.getShape(), memrefTy.getElementType());
   auto floatInit = mlir::DenseElementsAttr::get(tensorType, floatValue);
-  auto alignment = builder.getIntegerAttr(builder.getI64Type(), 128);
+  auto floatType = IntegerType::get(builder.getContext(), 64);
```
Comment (rengolin) [inline tpp-run/MLIRBench.cpp:362]: Float type? This is the alignment as an integer.

Also, why the change from `getI64Type()` to `IntegerType::get(64)`? They yield exactly the same result.

File: tpp-run/MLIRBench.cpp [hunk 7]
```diff
@@ -314,8 +338,12 @@ llvm::StringRef MLIRBench::createGlobal(MemRefType type) {
   static unsigned order = 0;
 
   // TODO: Use some random initialiser
-  APFloat floatValue = APFloat(1.0F);
-
+  auto floatValue = APFloat(1.0);
```
Comment (rengolin) [inline tpp-run/MLIRBench.cpp:341]: I think you can just do:
```
auto floatValue = APFloat(APFloat::BFloat(), 1.0);
```
or something

File: tpp-run/MLIRBench.cpp [hunk 8]
```diff
@@ -238,10 +252,20 @@ LogicalResult MLIRBench::printMemRef(mlir::Value memRef) {
...
+
+    minusOne = builder.create<arith::ConstantFloatOp>(
+        unkLoc, vectorFloatValue, FloatType::getBF16(builder.getContext()));
+  } else {
+    minusOne = builder.create<arith::ConstantFloatOp>(unkLoc, vectorFloatValue,
+                                                      builder.getF32Type());
+  }
 
-  // Loop through memref, transfer each dim to vector
+  // Vector undefined value
```
Comment (rengolin) [inline tpp-run/MLIRBench.cpp:268]: By "Vector undefined value" here I meant: When creating a vector, we can't leave `undef`, so we leave `-1.0` as the value that hasn't been filled by a memref read.

The following code is still looping through the memref...

File: tpp-run/MLIRBench.cpp (hunk 6 above)
Comment (KavithaTipturMadhu) [reply tpp-run/MLIRBench.cpp:362]: This is a stray change I forgot to revert locally. Will revert this, my bad, sorry.

File: tpp-run/MLIRBench.cpp [hunk 9]
```diff
@@ -238,8 +248,18 @@ LogicalResult MLIRBench::printMemRef(mlir::Value memRef) {
 
   // Vector undefined value
   APFloat vectorFloatValue = APFloat(-1.0F);
-  auto minusOne = builder.create<arith::ConstantFloatOp>(
-      unkLoc, vectorFloatValue, builder.getF32Type());
+  Value minusOne;
+  if (outputType.getElementType().isBF16()) {
+    bool Ignored;
```
Comment (chelini) [inline tpp-run/MLIRBench.cpp:253]: Ignored -> ignored (use camelCase).

File: tpp-run/MLIRBench.cpp [hunk 10]
```diff
@@ -314,8 +334,13 @@ llvm::StringRef MLIRBench::createGlobal(MemRefType type) {
   static unsigned order = 0;
 
   // TODO: Use some random initialiser
-  APFloat floatValue = APFloat(1.0F);
+  auto floatValue = APFloat(1.0F);
 
+  if (type.getElementType().isBF16()) {
+    bool Ignored;
```
Comment (chelini) [inline tpp-run/MLIRBench.cpp:340]: same.

**Review by chelini — APPROVED** [review]

File: test/BF16/mlp-all-bf16-tpprun.mlir [hunk 11]
```diff
@@ -0,0 +1,54 @@
+// RUN: tpp-opt %s -convert-check-to-loops -linalg-ext-to-loops -convert-linalg-to-loops -convert-tpp-to-xsmm -convert-xsmm-to-func -arith-expand -convert-math-to-llvm  -convert-vector-to-scf -convert-scf-to-cf -lower-affine  -reconcile-unrealized-casts |\
```
Comment (rengolin) [inline test/BF16/mlp-all-bf16-tpprun.mlir:1]: This test could be a benchmark (in `test/Benchmark`) and we should add a C/TPP version of it later, then you don't need to name it "tpprun".

File: test/BF16/mlp-all-bf16.mlir [hunk 12]
```diff
@@ -41,7 +41,9 @@ module @predict_function {
     linalg.fill ins(%c1:bf16) outs(%interim1: memref<128x512xbf16>)
     %threshold = arith.constant 0.75:bf16
     check.expect_almost_eq(%interim1, %arg12, %threshold): memref<128x512xbf16>, memref<128x512xbf16>, bf16
- 
+    memref.dealloc %interim1: memref<128x512xbf16>
```
Comment (rengolin) [inline test/BF16/mlp-all-bf16.mlir:44]: Do we need the two tests? If the kernels are the same, and the other is testing tpp-run as well, than this test can be removed.

**Review by rengolin — COMMENTED** [review]
Some comments but looks good, thanks!

File: test/BF16/mlp-all-bf16-tpprun.mlir (hunk 11 above)
Comment (KavithaTipturMadhu) [reply test/BF16/mlp-all-bf16-tpprun.mlir:1]: I'll move it to test/Benchmark in a different patch. Right now, we're categorizing all BF16 tests into one directory.

File: test/BF16/mlp-all-bf16.mlir (hunk 12 above)
Comment (KavithaTipturMadhu) [reply test/BF16/mlp-all-bf16.mlir:44]: You're right, I'll remove this one. It's doing the same thing as tpprun test except initializations are explicit. We can reintroduce it after random initialization is implemented in tpp-run.

File: test/BF16/mlp-all-bf16-tpprun.mlir (hunk 11 above)
Comment (rengolin) [reply test/BF16/mlp-all-bf16-tpprun.mlir:1]: Good point!

**Review by rengolin — APPROVED** [review]
